Seeds segmentation

Britefury opened the pull request:
I've knocked together an implementation of the SEEDS algorithm - or perhaps - an attempt at one. It produces colour based segmentations much like SLIC, except that it does not show the regular grid structure that is seen in the literature.
Much of my reference was based on [David Stutz's implementation](https://github.com/davidstutz/seeds-revised) and to a lesser extent an [implementation](https://github.com/Itseez/opencv_contrib/blob/master/modules/ximgproc/src/seeds.cpp) contributed to OpenCV. I hope my implementation is correct; I apologise if it isn't. If it is - with any luck - it can be used as a starting point for something better.

Britefury committed:
Implementation of SEEDS superpixel segmentation algorithm.

Britefury committed:
Added SEEDS example to plot_segmentation.

JDWarner commented:
Thanks for your contribution, @Britefury! This looks really interesting.

Our Travis continuous integration build is having trouble because it uses Bento. The new Cython file needs to be added to the `bento.info` file at package root.

I'll do a full review once the CI builds are happy.

Britefury committed:
Added SEEDS to bento.info

Commit:
dtype fix and docstring fixes.

Britefury committed:
dtype fixes.

Commit:
Further dtype fixes.

Commit:
type fix again.

Commit:
dtype fixes yet again.

Commit:
Added unit tests for SEEDS segmentation.

Commit:
Python 3 fix.

Commit:
Python 3 fix; xrange -> range.

Britefury commented:
Whew. That took a while to get right! :)

jni commented on skimage/segmentation/_seeds.pyx:
```diff
@@ -0,0 +1,725 @@
+        self.w = metrics.w
+        self.block_h = metrics.block_h * 2
+        self.block_w = metrics.block_w * 2
+        self.n_blocks_y = metrics.n_blocks_y // 2
```
These need to be updated to use r/c instead of y/x. See our [coordinate conventions](http://scikit-image.org/docs/dev/user_guide/numpy_images.html#coordinate-conventions) guide. (This is a new standard, which is why you might see some xyz in the code still. But we aim to transition to r/c completely.)

Britefury replied:
Fixed.

jni commented on skimage/segmentation/_seeds.pyx:
```diff
@@ -0,0 +1,725 @@
+
+        Parameters
+        ----------
+        y, x: the block co-ordinates
```
The convention here is:

```
param_name : param_type
    param_description
```

jni replied:
See the [NumPy docstring conventions](https://github.com/numpy/numpy/blob/master/doc/HOWTO_DOCUMENT.rst.txt)

Britefury replied:
Done.

jni commented on skimage/segmentation/_seeds.pyx:
```diff
@@ -0,0 +1,725 @@
+
+
+@cython.boundscheck(False)
+def upscale_labels_by_2(int[:,:] labels_2d, int h, int w):
```
I really don't understand this function, but it seems to me that it could be replaced by a call to `ndimage.zoom`...

jni commented on skimage/segmentation/_seeds.pyx:
```diff
@@ -0,0 +1,725 @@
+        up_labels_2d[:h2:2,-1] = labels_2d[:,-1]
+        up_labels_2d[1:h2:2,-1] = labels_2d[:,-1]
+
+
```
PEP8: remove extra blank line

Britefury replied:
Done.

jni commented on skimage/segmentation/_seeds.pyx:
```diff
@@ -0,0 +1,725 @@
+                block_w = metrics.last_block_w
+            for j in range(block_h):
+                for i in range(block_w):
+                    n = index_2d[y*metrics.block_h+j,x*metrics.block_w+i]
```
You probably mean to use just `block_h` and `block_w` rather than `metrics.block_?` here? If so, please add a test that will catch this error in the future.

Britefury replied:
No, I meant to use `metrics.block_h`/`w`. `metrics.block_h` contains the height of a block, where `block_h` contains either that or the height of a block in the bottom row if the image height is shorter than an exact multiple of `metrics.block_h`, so `y * metrics.block_h` is the offset used to move `j` into the current block.

jni commented on skimage/segmentation/_seeds.pyx:
```diff
@@ -0,0 +1,725 @@
+    Check if re-labelling a block or pixel will break a labelled region
+    into disjoint sub-regions.
+
+    An explanation; assume that we are relabelling a pixel to match that
```
LOL, remove "An explanation" =)

Britefury replied:
Done.

jni commented on skimage/segmentation/_seeds.pyx:
```diff
@@ -0,0 +1,725 @@
+    The pixel behind/below it is `b0`.
+    The pairs `a0`/`a1` and `c0`/`c1` represent pixels either side, e.g.
+    `a0` represents the pixel to the left and behind, while `a1` represents
+    the pixel to the right. `c0`/`c1` represent pixels to the right.
```
This made no sense to me... Perhaps some ascii art is warranted... See for example the documentation for [label](http://scikit-image.org/docs/dev/api/skimage.measure.html#label)

Britefury replied:
Ascii art added as requested.

jni commented on skimage/segmentation/_seeds.pyx:
```diff
@@ -0,0 +1,725 @@
+    Returns
+    -------
+    A boolean indicating if the proposed relabelling would cause a break.
+    """
```
Think about how your calls will read when designing an API. You want it to be as close as possible to English. In this case, for example, you could call it `_relabel_is_safe`, or you could reverse the boolean result and call it `_relabel_will_disconnect`.

Britefury replied:
Now using `_relabel_will_disconnect`.

jni commented on skimage/segmentation/_seeds.pyx:
```diff
@@ -0,0 +1,725 @@
+    b0 = labels_2d[y+1,x]   if y<(h-1)   else -1
+    b1 = labels_2d[y,x]
+    c0 = labels_2d[y+1,x+1]   if y<(h-1) and x<(w-1)   else -1
+    c1 = labels_2d[y,x+1]   if x>0 and x<(w-1)   else -1
```
This is a huge mess of PEP8 noncompliance...

Britefury replied:
Fixed.

jni commented:
@Britefury thanks for submitting this! I hadn't heard of SEEDS and it looks to me like a great algorithm, worthy of inclusion in scikit-image. 

That's the good news. The bad news is that I think the implementation will need a lot of refinement before getting merged. Functions like `refine_blocks`, `refine_pixels`, and `seeds` are huge monoliths that will be very difficult to maintain in the future.

A lot of the code can be simplified using scipy and numpy functions. In fact, from what I've seen, the implementation can be made n-dimensional *and* the code simplified, to boot. For example, instead of your 2D histogram downsampling function, you can use `numpy.add.reduceat` to downsample a histogram of any dimension:

```python
import numpy as np
# 2D histogram for 9 x 16 image, 3 bins
h = np.random.rand(9, 16, 3)
hs = h
for ax in range(hs.ndim - 1):
    hs = np.add.reduceat(hs, np.arange(0, hs.shape[ax] - 1, 2), axis=ax)
```

`hs` will hold the downsampled histogram.

jni commented:
Also, we tend to prefer a functional approaches here... The `LevelMetrics` class doesn't seem to add much functionality, so e.g. a dictionary might be more appropriate.

We're definitely willing to iterate with you on this! I don't want to discourage you! But it's going to take a few tries to get right, I think.

Britefury committed:
y,x -> r,c convention compliance.
PEP8 fixes.
_relabel_check* functions renamed to _relabel_will_disconnect*.
Improved docstring of _relabel_check.

Britefury committed:
Merge changes from different clones.

Britefury committed:
Some more y,x -> r,c convention fixes, along with a docstring style fix.

Britefury commented:
I benchmarked `upscale_labels_by_2` vs `ndimage.zoom`. Given a 256x256 input image, dtype=int, `upscale_labels_by_2` takes 1.34ms while `ndimage.zoom` takes 4.3ms.
Is performance an adequate reason to keep `upscale_labels_by_2` around?

ahojnnes commented:
What's the overall runtime for a 256x256 input image?

JDWarner commented:
We also need to know how they scale. 256, 512, and 1024 square benchmarks for both would be ideal.

Britefury commented:
1024x1024: 14.5s
512x512: 3.75s
256x256: 1.22s

ahojnnes commented:
Comparing that to the 1-5ms that the upscaling takes, it does not seem worth to replicate that code. We should just go with the ndimage.zoom function?

JDWarner commented:
How does ndi.zoom scale with input size?

Commit:
Replaced hand-rolled upscale_labels_by_2 cython function with a Python function that uses numpy.repeat, also netting a speed increase.

Britefury commented:
Have found a compromise: it seems that applying `numpy.repeat` a couple of times is even faster than the hand-rolled solution. Part of the original `upscale_labels_by_2` implementation remains in order to repeat the last row/column when needed. `upscale_labels_by_2` has been moved from `_seeds.pyx` to `seeds_superpixels.py`.

sciunto commented:
This feature looks cool. What's the status of this PR guys?

alexandrejaguar commented:
Hi @scikit-image/core,
is this something we still want in? Thanks!

jni commented:
@alexandrejaguar if we can make it ND I'm all for it! =) Thank you for all your work on picking up dropped PRs, it's wonderful to see!

hmaarrfk commented:
> @alexandrejaguar if we can make it ND I'm all for it! =) Thank you for all your work on picking up dropped PRs, it's wonderful to see!

I think this ND requirements really need to be discussed.

jni commented:
> I think this ND requirements really need to be discussed.

You are right. I am firm in my assessment, though, but could be overruled... Assuming we get governance #3585 through soon! =\ 

Please see recent commits in #3544 for work I'm doing getting our transforms to support 3D/nD data. If we aim to be "maximally" nD compliant, which I think we do (as nD is important to a growing proportion of our users), then merging before code is nD amounts to taking on technical debt, of which we already have too much imho.

Part of this can be fixed with labeling function names as `_2d()`. That at least makes clear to a user interested in our 3D processing that they shouldn't even look at those functions. But imho in many cases, including this one, the code can be written as nD up-front with not too much extra effort, and that should be done before merging, because then it remains on a to-do list.

If we had users banging on the door asking for SEEDS in skimage, I'd be inclined to be more flexible, but 2D SEEDS is already [available](https://ddokkddokk.tistory.com/18) in [OpenCV](https://docs.opencv.org/3.0-beta/modules/ximgproc/doc/superpixels.html). What can we bring to the table? As someone who teaches git and open source contribution using GitHub on a regular basis, I very much sympathise with the “all contributions are good, let's have more of them!” mindset, *but* for a project like scikit-image, it needs to be balanced with maintenance cost. We should push to get contributions in *if* they meet a certain level of utility. imho, nD SEEDS would meet that threshold, while 2D SEEDS is borderline.

hmaarrfk commented:
Gah. Sorry @jni, didn't see this when I sent out the email on the smiling list, should we move the discussion to an issue on GitHub?

jni commented:
The smiling list! We should rename it that. =P

I think this is worth discussing in both places, as they reach different audiences. Your point about the API is a very good one.

hmaarrfk commented on skimage/segmentation/_seeds.pyx:
```diff
@@ -0,0 +1,709 @@
+    cdef int[:,:] up_labels_2d = np.zeros((metrics.h, metrics.w), dtype=np.int32)
+    cdef int c, r, i, j, block_w, block_h
+
+    for r in range(metrics.n_blocks_r):
```
can you release the GIL here?

hmaarrfk commented on skimage/segmentation/_seeds.pyx:
```diff
@@ -0,0 +1,709 @@
+    cdef int[:,:,:] hist_2d = np.zeros((metrics.n_blocks_r, metrics.n_blocks_c,
+                                        hist_size), dtype=np.int32)
+    cdef int c, r, i, j, block_h, block_w, n
+    for r in range(metrics.n_blocks_r):
```
can ou release the GIL here?

hmaarrfk commented on skimage/segmentation/_seeds.pyx:
```diff
@@ -0,0 +1,709 @@
+    cdef int c, r, i
+    cdef int rem_r = hist_2d.shape[0] & 1
+    cdef int rem_c = hist_2d.shape[1] & 1
+    for r in range(h2):
```
can you release the GIL here?

hmaarrfk commented on skimage/segmentation/_seeds.pyx:
```diff
@@ -0,0 +1,709 @@
+    cdef double l0n = 0.0, l1n = 0.0, bn = 0.0, l0int=0.0, l1int=0.0;
+    cdef int i
+
+    if label_0 != label_1:
```
can you release the GIL here?

hmaarrfk commented on skimage/segmentation/_seeds.pyx:
```diff
@@ -0,0 +1,709 @@
+
+@cython.boundscheck(False)
+cdef bint _relabel_will_disconnect(int a0, int a1, int b0, int b1,
+                                   int c0, int c1):
```
should be declared `nogil`

hmaarrfk commented on skimage/segmentation/_seeds.pyx:
```diff
@@ -0,0 +1,709 @@
+
+@cython.boundscheck(False)
+cdef bint relabel_will_disconnect_above(int[:,:] labels_2d, int r, int c,
+                                        int h, int w):
```
are these internal function? if so, these little C helper routines should be declared `nogil`

hmaarrfk commented on skimage/segmentation/_seeds.pyx:
```diff
@@ -0,0 +1,709 @@
+
+
+@cython.boundscheck(False)
+cdef bint relabel_will_disconnect_above(int[:,:] labels_2d, int r, int c,
```
should `lablels_2d` be declared contiguous? `int [:, ::1] labels_2d`????

hmaarrfk commented on skimage/segmentation/_seeds.pyx:
```diff
@@ -0,0 +1,709 @@
+    cdef int refine_it, r, c
+
+    # For each refinement iteration
+    for refine_it in range(refine_iters):
```
should be `nogil`

hmaarrfk commented on skimage/segmentation/_seeds.pyx:
```diff
@@ -0,0 +1,709 @@
+    cdef int refine_it, r, c
+
+    # For each refinement iteration
+    for refine_it in range(refine_iters):
```
nogil?

hmaarrfk commented on skimage/segmentation/seeds_superpixels.py:
```diff
@@ -0,0 +1,273 @@
+           Proceedings of the European Conference on Computer Vision,
+           pages 13-26, 2012.
+    """
+    assert n_levels >= 1
```
These assert statements should be replaced by real `if` statements that raising meaningful error messages

hmaarrfk commented on skimage/segmentation/seeds_superpixels.py:
```diff
@@ -0,0 +1,273 @@
+    seed_h = max(int(block_finest_h_f+0.5), 1)
+
+    # Get an index image; determine what to do.
+    if len(input_2d.shape) == 3 and input_2d.shape[2] == 3:
```
We've been working against this kind of "magic". Not too sure what to do in this case, but I'm just flagging it for now. We have been going with the `multichannel` keyword argument asking users to explicitely label their images.

hmaarrfk commented on skimage/segmentation/seeds_superpixels.py:
```diff
@@ -0,0 +1,273 @@
+                                              iter=50)
+        index_2d = index_1d.reshape(input_2d.shape[:2])
+    elif len(input_2d.shape) == 2:
+        if input_2d.dtype == float or input_2d.dtype == np.float32:
```
you can check `issubtype`

hmaarrfk commented on skimage/segmentation/tests/test_seeds.py:
```diff
@@ -0,0 +1,32 @@
+    assert_equal(labels[:16, 16:], 2)
+    assert_equal(labels[16:, 16:], 4)
+
+if __name__ == '__main__':
```
this is no longer necessary. Pytest will just find the test above.

hmaarrfk commented on skimage/segmentation/tests/test_seeds.py:
```diff
@@ -0,0 +1,32 @@
+
+
+@test_parallel()
+def test_color_2d():
```
You take in so many different parameters, and check for all of the. You should really test many more keyword argument and "bad input" cases.


Especially since you take in and special case many different types of dtypes.

hmaarrfk reviewed (commented):
Didn't really dwelve too much into the code, but at first glance:
1. Try to release the GIL. Not sure if it is always possible.
2. If you are special casing dtypes, then you should test for all the dtypes you consider.

matthew-brett commented:
We have changed the base branch for this PR from the `main` branch to the `pre-megamove-integration` branch.

## Summary - what this means for your PR

You can continue to work on your PR, now based on `pre-megamove-integration`.  You should update your branch to the current state of `pre-megamove-integration` (this is what the Github "Update branch" button does).   Please **do not merge the scikit-image `main` branch into your PR branch**.

Once we've merged your PR into the `pre-megamove-integration` branch, we'll update your changes for Megamove, and merge into the `main` branch.

In addition - our automated script detected that your branch had merge conflicts with the pre-Megamove branch. Please make sure to merge in the `pre-megamove-integration` branch, and resolve the merge conflicts in the way that you want.

Read on for the Megamove rationale, and more detailed instructions.

<details>

## Rationale

We've changed the base branch to make it easier and smoother to continue to work on your PR, while we work through a rather large change in the directory structure of the repository - a change that we have called *Megamove*.

See [the Megamove explanation issue](https://github.com/scikit-image/scikit-image/issues/8232) for more detail.

Changing the base branch means you do not see merge conflicts related to the *Megamove* change, and can continue to work with the repository in its original state, without worrying about porting those changes across to the new post-Megamove structure.

As you'll see in the linked issue, you can continue to work on your PR as before, and you can continue to use the Github GUI to "Update branch" - this updates from the new `pre-megamove-integration` branch.   However, there is one caveat: please, do not try and merge the `scikit-image` `main` branch into this one, but instead, keep track of the `pre-megamove-integration` branch.  The issue explains more.

To emphasize again, **do not try and merge the scikit-image `main` branch**.  You will get Megamove related merge conflicts that are rather tricky to resolve correctly.

When we've approved your PR, we'll merge into the `pre-megamove-integration` branch, and then process it so we can merge it into the `main` branch.   We'll let you know when we've merged your PR changes into `main`.

</details>

Any questions, please reply here.
